pdf/pdfselection.cpp

```diff
@@ -0,0 +1,439 @@
+
+    d->stop = QPointF(box.x() + box.width(), box.y() + box.height() / 2.);
+    emit stopChanged();
+}
+
+void PDFSelection::selectAt(const QPointF &pt)
+{
+    int iPage, iBox;
+    d->textBoxId(pt, PDFSelection::Private::At, &iPage, &iBox);
+    if (iPage < 0 || iBox < 0) {
+        beginResetModel();
+        d->iPageStart = d->iPageStop = -1;
+        d->iBoxStart = d->iBoxStop = -1;
+        endResetModel();
+        emit countChanged();
+        emit textChanged();
+        return;
+    }
+
+    const PDFDocument::TextList &boxes = d->canvas->document()->textBoxesAtPage(iPage);
+    QRectF box = boxes[iBox].first;
+    
+    beginResetModel();
+    d->iPageStart = -1;
+    d->iBoxStart = -1;
+    d->iPageStop = -1;
+    d->iBoxStop = -1;
+    endResetModel();
+    beginInsertRows(QModelIndex(), 0, 0);
+    d->iPageStart = iPage;
+    d->iBoxStart = iBox;
+    d->iPageStop = iPage;
+    d->iBoxStop = iBox;
+    endInsertRows();
+    emit countChanged();
+    emit textChanged();
+
+    d->start = QPointF(box.x(), box.y() + box.height() / 2.);
+    emit startChanged();
+    d->stop = QPointF(box.x() + box.width(), box.y() + box.height() / 2.);
+    emit stopChanged();
+}
+
+void PDFSelection::onLayoutChanged()
+{
+    emit dataChanged(createIndex(0, 0), createIndex(count() - 1, 0), QVector<int>{Rect});
+}
+
+bool PDFSelection::selectionAtPoint(const QPointF &pt) const
+{
+    if (d->iPageStart < 0 || d->iPageStop < 0 ||
+        d->iBoxStart < 0 || d->iBoxStop < 0)
+        return false;
+
+    if (!d->canvas)
+        return false;
+    PDFDocument *doc = d->canvas->document();
+    if (!doc)
+        return false;
+    
+    int i;
+    QPair<int, QRectF> at = d->canvas->pageAtPoint(pt);
+    if (at.first < 0)
+        return false;
+    QPointF ptRed {pt.x() / at.second.width(),
+            (pt.y() - at.second.y()) / at.second.height()};
+    const PDFDocument::TextList &boxes = doc->textBoxesAtPage(at.first);
+    for (i = ((at.first == d->iPageStart) ? d->iBoxStart : 0);
+         i < ((at.first == d->iPageStop) ? d->iBoxStop + 1 : boxes.length());
+         i++) {
+        if (boxes.value(i).first.contains(ptRed))
+            return true;
+    }
+    return false;
+}
+
+QString PDFSelection::text() const
+{
+    QString out;
+
+    if (d->iPageStart < 0 || d->iPageStop < 0 ||
+        d->iBoxStart < 0 || d->iBoxStop < 0)
+        return out;
+
+    if (!d->canvas)
+        return out;
+    PDFDocument *doc = d->canvas->document();
+    if (!doc)
+        return out;
+    
+    const PDFDocument::TextList &boxes = doc->textBoxesAtPage(d->iPageStart);
+    out = boxes.value(d->iBoxStart).second;
+    int i, j;
+    for (i = d->iPageStart; i <= d->iPageStop; i++) {
+        const PDFDocument::TextList &boxes = doc->textBoxesAtPage(i);
+        for (j = ((i == d->iPageStart) ? d->iBoxStart + 1 : 0);
+             j < ((i == d->iPageStop) ? d->iBoxStop + 1 : boxes.length());
+             j++) {
+            out += " ";
```

Comment:
I use, as you suggest, the routine hasSpaceAfter() from Poppler, and use a little trick here. Poppler objects are only accessible in pdfrenderthread.cpp. So the hasSpaceAfter() is lost when reaching pdfselection.cpp, only rectangle and text are stored. The trick consists in appending a space to the stored string when hasSpaceAfter() is true. The text is then reconstructed by concatenating all the string without adding space inbetween each. This trick may leave a trailing space on the selected text.